### PR TITLE
ACTIONX: Fix bug with matching wells + refactor

### PR DESCRIPTION
**Pull request by joakim-hove**
When evaluating an `ACTIONX` expression like:
```
ACTIONX
    WWCT > 0.75 /
/
```
the result should contain an overall thruthness[1] and the set of wells matching the condition. In this PR the bool and the set of wells is assemble together in a small class `Action::Result`. The main new functionality in this PR is that expressions like:
```
ACTIONX
    FWPR > 1000 AND  WWCT < 0.75 /
/ 
```
work correctly. Previously the first condition `FWPR > 1000` had no matching wells - and when the empty set of wells was intersected with the wells matching `WWCT < 0.75` the set of matching wells remained empty - this is fixed in this PR. In addition the PR introduces a `Action::` namespace, and even some documentation - everything is in the large third commit.

[1]: The overall truthness in an expression like this is whether *any* well satisfies the criteria.

### CODE REVIEW

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
I'll let this linger over the summer holiday; taking a look at this might be a healthy(?) exercise - as it should provide a gentle introduction to the ACTIONX machinery.

File: src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    return this->matching_wells.wells();
+}
+
+bool Result::scalar() const {
+    return matching_wells.scalar();
+}
+
+
+Result& Result::operator|=(const Result& other) {
+    this->m_result |= other.m_result;
```
Comment (bska) [inline src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp]: Bitwise 'or' is not defined for `bool`.  You need to say
```C++
this->m_result = this->m_result || other.m_result;
```

File: src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    this->matching_wells.add( other.matching_wells );
+    return *this;
+}
+
+Result& Result::operator&=(const Result& other) {
+    this->m_result &= other.m_result;
```
Comment (bska) [inline src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp]: Bitwise 'and' is not defined for `bool`.  You need to say
```C++
this->m_result = this->m_result && other.m_result;
```

File: src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp (hunk 1 above)
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp]: I believe you - of course - and will update, but the current code compiles without warning and works as intended, maybe that is just an undefined fallback to the whatever underlying integral type?

File: opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.hpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+
+
+class Result {
+public:
+    explicit Result(bool result_arg);
+    Result(bool result_arg, const std::vector<std::string>& wells);
+    Result(bool result_arg, const WellSet& wells);
+
+    bool result() const;
```
Comment (atgeirr) [inline opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.hpp]: I think the result() and scalar() members are quite confusing.

As I understand the PR description, the (or one) problem is dealing with expressions like `FWPR > 1000` which generate a result with an empty set of wells currently. The simple fix would then be to have such expressions generate a result object containing all wells. That way, a Result is "the set of wells for which the conditions hold" always.

A possible optimization would be to have a special flag for "all wells". I think that is what you have done here, but I do not understand why this state seems to be called "scalar". Furthermore I think calling wells() will not give the correct set of wells (all of them...) for such a Result.

The result() member is also confusing. I'll assume the member is supposed to be understood like this: If result() is `true`, then I contain all wells for which the conditions hold, if result() is false, then I contain all wells for which the conditions do not hold. Then again the wells() member is suspect, or at least hard to use: its meaning depends on the value of result(). I think a Result should always behave as if result() is true, if you need to negate it I would rather have a `not` operator doing that.

File: src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+
+WellSet& WellSet::intersect(const WellSet& other) {
+    if (this->m_scalar)
+        *this = other;
+    else {
+        auto well_iter = this->well_set.begin();
+        while (well_iter != this->well_set.end()) {
+            if (!other.contains(*well_iter))
+                well_iter = this->well_set.erase(well_iter);
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp]: Fixed

File: src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    return *this;
+}
+
+WellSet& WellSet::add(const WellSet& other) {
+    if (this->m_scalar)
+        *this = other;
+    else {
+        for (auto& well : other.well_set)
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp]: Done

File: src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+Result::Result(bool result_arg, const WellSet& wells) :
+    m_result(result_arg),
+    matching_wells(wells)
+{}
+
+bool Result::result() const {
+    return this->m_result;
+}
+
+std::vector<std::string> Result::wells() const {
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp]: Yes - we could return a `const` reference to the underlying set; the reason I converted to `std::vector` was that *everyone likes a vector` - the `std::set` is slightly less standard - as I see it. No strong feelings though.

File: src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    return *this;
+}
+
+Result& Result::operator&=(const Result& other) {
+    this->m_result = this->m_result && other.m_result;
+    this->matching_wells.intersect( other.matching_wells );
+    return *this;
+}
+
+void Result::assign(double value) {
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionResult.cpp]: Fixed

**Comment by joakim-hove** [comment]
Thank you for the review - I dare say you have misunderstood the code at several points, but that just amounts to saying the code was less than clear. I have now updated it - personally I think it is much clearer now - hopefully you agree.

The main problem was to differentiate between a result which has no wells - because that is not relevant:
```
FPR > 100
```
and a result of *well type* - with zero matching wells:
```
WWCT > 1.25
```
that is now solved by using pointer semantics internally on the `WellSet` - where the first example will have a `nullptr`, whereas the last example will point to an empty set - that has done away with the `scalar()` uglyness completely. 

Have also implemented `operator bool()` and dropped the `result()` method, finally some code has been rewritten to enable removing the `assign(bool)` method - we don't like setters do we?

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
@atgeirr - seems you are back in action; cool if you give this another round.

**Comment by bska** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
OK - I'll merge this on thursday.

**Comment by joakim-hove** [comment]
jenkins build this please

**Pull request merged**